chainer/training/extensions/snapshots/writer.py

```diff
@@ -0,0 +1,223 @@
+import threading
+
+from chainer.training.extensions.snapshots import util
+
+
+class Writer(object):
+    """Base class of snapshot writer.
+
+    :class:`Snapshot` invokes this :class:`Writer` object's :func:`__call__`
+    everytime if this process is in charge to take a snapshot.
+    This class determine how to invoke the actual saving function. The most
+    simple way is just passing the arguments to the saving function.
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self, filename, outdir, target):
+        """Invokes the actual snapshot function.
+
+        This method is invoked every time when this process is in charge to
+        take a snapshot. All arguments are passed from :class:`Snapshot`
+        object.
+
+        Args:
+            filename (str): Name of the file into which the serialized target
+                is saved. It is already formated string.
+            outdir (str): Output directory. Passed by `trainer.out`.
+            target (dict): Serialized object which will be saved.
+        """
+        pass
+
+    def finalize(self):
+        """Finalize the writer.
+
+        Like extensions in :class:`Trainer`, this method is invoked at the
+        end of training.
+        """
+        pass
+
+    def save(self, filename, outdir, target, savefun, **kwds):
+        prefix = 'tmp' + filename
+        fd, tmppath = tempfile.mkstemp(prefix=prefix, dir=outdir)
+        try:
+            savefun(tmppath, target, **kwds)
+        except Exception:
+            os.close(fd)
+            os.remove(tmppath)
+            raise
+        os.close(fd)
+        shutil.move(tmppath, os.path.join(outdir, filename))
+
+
+class SimpleWriter(Writer):
+    """The most simple writer.
+
+    This class just passes the arguments to the actual saving function.
+
+    Args:
+        savefun: Callable object. Must have same interface to `Writer.__call__`
+            method.
+    """
+
+    def __init__(self, savefun=util.save_npz, **kwds):
+        self._savefun = savefun
+        self._kwds = kwds
+
+    def __call__(self, filename, outdir, target):
+        self.save(filename, outdir, target, self._savefun, **self._kwds)
+
+
+class ThreadWriter(Writer):
+    """Writer that uses a thread.
+
+    This class creates a thread that invokes the actual saving function.
+    A new thread is created every time `__call__` is invoked.
+
+    Args:
+        savefun: Callable object. Must have same interface to `Writer.__call__`
+            method. Also must need to be able to passed to a thread.
+    """
+
+    def __init__(self, savefun=util.save_npz, **kwds):
+        self._savefun = savefun
+        self._kwds = kwds
+        self._flag = False
+
+    def __call__(self, filename, outdir, target):
+        self._thread = threading.Thread(
+            target=self.save,
+            args=(filename, outdir, target, self._savefun),
+            kwargs=self._kwds)
+        self._thread.start()
+        self._flag = True
```

Comment:
The previous write operation is not guaranteed to be completed, which can lead to unstable behavior if `filename` is not changed between calls. How about joining the previous thread?